File: VIF_git/GSDS-Qual/algorithms/sort/heap_sort.cpp
#include <iostream>
using namespace std;

// To heapify a subtree rooted with node i
// which is an index in arr[].
// n is size of heap
void maxHeapify(int arr[], int N, int i)
{

    int largest = i;
    int l = 2 * i + 1;
    int r = 2 * i + 2;

    if (l < N && arr[l] > arr[largest])
        largest = l;

    if (r < N && arr[r] > arr[largest])
        largest = r;

    // If largest is not root
    if (largest != i) {
        swap(arr[i], arr[largest]);

        maxHeapify(arr, N, largest);
    }
}

void minHeapify(int arr[], int N, int i)
{

    int smallest = i;
    int l = 2 * i + 1;
    int r = 2 * i + 2;

    if (l < N && arr[l] < arr[smallest])
        smallest = l;

    if (r < N && arr[r] < arr[smallest])
        smallest = r;

    // If smallest is not root
    if (smallest != i) {
        swap(arr[i], arr[smallest]);

        minHeapify(arr, N, smallest);
    }
}

void heapSortAscending(int arr[], int N)
{

    // Build heap (rearrange array)
    for (int i = N / 2 - 1; i >= 0; i--)
        maxHeapify(arr, N, i);

    // One by one extract an element
    // from heap
    for (int i = N - 1; i > 0; i--) {

        // Move current root to end
        swap(arr[0], arr[i]);

        // call max heapify on the reduced heap
        maxHeapify(arr, i, 0);
    }
}

void heapSortDescending(int arr[], int N)
{

    // Build heap (rearrange array)
    for (int i = N / 2 - 1; i >= 0; i--)
        minHeapify(arr, N, i);

    // One by one extract an element
    // from heap
    for (int i = N - 1; i > 0; i--) {

        // Move current root to end
        swap(arr[0], arr[i]);

        // call min heapify on the reduced heap
        minHeapify(arr, i, 0);
    }
}

// A utility function to print array of size n
void printArray(int arr[], int N)
{
    for (int i = 0; i < N; ++i)
        cout << arr[i] << " ";
    cout << "\n";
}

// Driver's code
int main()
{
    int arr[] = { 12, 11, 13, 5, 6, 7 };
    int N = sizeof(arr) / sizeof(arr[0]);

    heapSortAscending(arr, N);

    cout << "Sorted array Ascending is \n";
    printArray(arr, N);

    heapSortDescending(arr, N);

    cout << "Sorted array Descending is \n";
    printArray(arr, N);
}

File: VIF_git/GSDS-Qual/algorithms/sort/heap_sort.py
def maxHeapify(arr, N, i):
    largest = i  # Initialize largest as root
    l = 2 * i + 1     # left = 2*i + 1
    r = 2 * i + 2     # right = 2*i + 2

    if l < N and arr[largest] < arr[l]:
        largest = l

    if r < N and arr[largest] < arr[r]:
        largest = r

    # Change root, if needed
    if largest != i:
        arr[i], arr[largest] = arr[largest], arr[i]  # swap

        # Heapify the root.
        maxHeapify(arr, N, largest)

def minHeapify(arr, N, i):
    smallest = i  # Initialize largest as root
    l = 2 * i + 1     # left = 2*i + 1
    r = 2 * i + 2     # right = 2*i + 2

    if l < N and arr[smallest] > arr[l]:
        smallest = l

    if r < N and arr[smallest] > arr[r]:
        smallest = r

    # Change root, if needed
    if smallest != i:
        arr[i], arr[smallest] = arr[smallest], arr[i]  # swap

        # Heapify the root.
        minHeapify(arr, N, smallest)

def heapSortAscending(arr):
    N = len(arr)

    # Build a maxheap.
    for i in range(N//2 - 1, -1, -1):
        maxHeapify(arr, N, i)

    # One by one extract elements
    for i in range(N-1, 0, -1):
        arr[i], arr[0] = arr[0], arr[i]  # swap
        maxHeapify(arr, i, 0)
        
def heapSortDescending(arr):
    N = len(arr)

    # Build a maxheap.
    for i in range(N//2 - 1, -1, -1):
        minHeapify(arr, N, i)

    # One by one extract elements
    for i in range(N-1, 0, -1):
        arr[i], arr[0] = arr[0], arr[i]  # swap
        minHeapify(arr, i, 0)


# Driver's code
if __name__ == '__main__':
    arr = [12, 11, 13, 5, 6, 7]

    heapSortAscending(arr)
    N = len(arr)

    print("Sorted array Ascending is")
    for i in range(N):
        print("%d" % arr[i], end=" ")

    print("")

    heapSortDescending(arr)

    print("Sorted array Descending is")
    for i in range(N):
        print("%d" % arr[i], end=" ")